File: client/python/mesh/mpc/consumer_filter.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#

from typing import Any, Dict

import mesh.log as log
from mesh.cause import MeshCode, Codeable, MeshException
from mesh.codec import Codec
from mesh.context import Mesh
from mesh.kinds import MeshFlag
from mesh.macro import spi, ServiceLoader
from mesh.mpc.digest import Digest
from mesh.mpc.filter import Filter, Invoker, Invocation, CONSUMER
from mesh.prsim import Metadata


@spi(name="consumer", pattern=CONSUMER, priority=0x80000000)
class ConsumerFilter(Filter):

    def invoke(self, invoker: Invoker, invocation: Invocation) -> Any:
        codec: Codec = ServiceLoader.load(Codec).get(MeshFlag.JSON.get_name())
        attachments: Dict[str, str] = invocation.get_parameters().get_attachments()
        for key, value in Mesh.context().get_attachments().items():
            attachments[key] = value
        attachments[Metadata.MESH_TRACE_ID.key()] = Mesh.context().get_trace_id()
        attachments[Metadata.MESH_SPAN_ID.key()] = Mesh.context().get_span_id()
        attachments[Metadata.MESH_TIMESTAMP.key()] = f"{Mesh.context().get_timestamp()}"
        attachments[Metadata.MESH_RUN_MODE.key()] = f"{Mesh.context().get_run_mode().value}"
        attachments[Metadata.MESH_URN.key()] = Mesh.context().get_urn()
        attachments[Metadata.MESH_CONSUMER.key()] = codec.encode_string(Mesh.context().get_consumer())
        attachments[Metadata.MESH_PROVIDER.key()] = codec.encode_string(Mesh.context().get_provider())

        digest = Digest()
        try:
            ret = invoker.run(invocation)
            digest.write("C", MeshCode.SUCCESS.get_code())
            return ret
        except BaseException as e:
            if isinstance(e, Codeable):
                digest.write("C", e.get_code())
            else:
                digest.write("C", MeshCode.SYSTEM_ERROR.get_code())
            if isinstance(e, MeshException):
                log.error(f"{digest.trace_id},{Mesh.context().get_urn()},{e.get_message()}")
            raise e

File: client/java/src/main/java/com/be/mesh/client/mpc/ConsumerFilter.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.mpc;

import com.be.mesh.client.annotate.SPI;
import com.be.mesh.client.cause.MeshException;
import com.be.mesh.client.prsim.Codeable;
import com.be.mesh.client.prsim.Context;
import com.be.mesh.client.struct.Cause;
import com.be.mesh.client.struct.Digest;
import lombok.extern.slf4j.Slf4j;

import java.util.Map;

/**
 * @author coyzeng@gmail.com
 */
@Slf4j
@SPI(value = "consumer", pattern = Filter.CONSUMER, priority = Integer.MIN_VALUE)
public class ConsumerFilter implements Filter {

    @Override
    public Object invoke(Invoker<?> invoker, Invocation invocation) throws Throwable {
        Codec codec = ServiceLoader.load(Codec.class).get(MeshFlag.JSON.getName());
        Map<String, String> attachments = invocation.getParameters().attachments();
        attachments.putAll(Mesh.context().getAttachments());
        Context.Metadata.MESH_TRACE_ID.set(attachments, Mesh.context().getTraceId());
        Context.Metadata.MESH_SPAN_ID.set(attachments, Mesh.context().getSpanId());
        Context.Metadata.MESH_TIMESTAMP.set(attachments, String.valueOf(Mesh.context().getTimestamp()));
        Context.Metadata.MESH_RUN_MODE.set(attachments, String.valueOf(Mesh.context().getRunMode().getMode()));
        Context.Metadata.MESH_URN.set(attachments, String.valueOf(Mesh.context().getUrn()));
        Context.Metadata.MESH_CONSUMER.set(attachments, codec.encodeString(Mesh.context().getConsumer()));
        Context.Metadata.MESH_PROVIDER.set(attachments, codec.encodeString(Mesh.context().getProvider()));

        Digest digest = new Digest();

        try {
            // Refresh slf4j MDC
            digest.mdc();
            Object ret = invoker.invoke(invocation);
            digest.print("C", MeshCode.SUCCESS.getCode());
            return ret;
        } catch (Throwable e) {
            Cause.POS.ifPresent(pos -> log.error("MPC invoke fault at {} with {}", pos, digest.getTraceId()));
            if (e instanceof Codeable) {
                digest.print("C", ((Codeable) e).getCode());
            } else {
                digest.print("C", MeshCode.SYSTEM_ERROR.getCode());
            }
            if (e instanceof MeshException) {
                log.error("{},{},{}", digest.getTraceId(), Mesh.context().getUrn(), ((MeshException) e).getRootMessage());
            }
            throw e;
        }
    }

}
